### ISSUE TITLE
ECONNREFUSED not captured

### ISSUE THREAD

**Issue by gabmontes** [sergi/jsftp#210]
Run across this issue when the remote server was down. Code to reproduce is simple:

``` js
const JSFtp = require('jsftp')

const ftp = new JSFtp({
  host: '127.0.0.1', port: 3333,
  user: 'user', pass: '1234'
})

ftp.ls('.', function(err, res) {
  console.log(res)
})
```

That code, when executed with the ftp server down, will throw as this while executing the list command:

``` bash
$ DEBUG=jsftp:* node jsftptest.js
  jsftp:command feat +0ms
  jsftp:command feat +3ms
events.js:160
      throw er; // Unhandled 'error' event
      ^

Error: connect ECONNREFUSED 127.0.0.1:3333
    at Object.exports._errnoException (util.js:1026:11)
    at exports._exceptionWithHostPort (util.js:1049:20)
    at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1085:14)
```

The expected behavior would be to receive the error in the callback `err` param. What do you think, guys? Consider the same issue is showing up with `ftp.put` and other commands, but `ls` looked like the simplest example.

I'm using jsftp@1.5.3.

**Update**

After more research and playing with domains to capture the error, I found the exceptions is thrown in the context of `new JSFtp` and not on the command itself (i.e. `ls`). Hope this helps to track down and solve the issue!

**Comment by davidjkrause** [comment]
Similar uncaught exception occurs if the hostname provided fails DNS lookup (ENOTFOUND).

I'm not an expert on JS or Node, but I've found at least one way to avoid uncaught exceptions and get the errors routed back to my application.

In this function in jsftp.js:

```
Ftp.prototype._createSocket = function(port, host, firstAction) { ...
```

I added Event emitter handlers for 'error' and 'lookup' on the socket returned from Net.createConnection, and then add further handlers for those in my application.

jsftp.js patch is something like:

```
  this.socket.on('connect', this.reemit('connect'));
  this.socket.on('timeout', this.reemit('timeout'));
+  this.socket.on('error', this.reemit('error'));
+  this.socket.on('lookup', this.reemit('lookup'));

  this.pipeline = combine(this.socket, this.resParser);
```

And in my code, I have:

```
ftpConn = new JSFTP { ... }

ftpConn.on('error', (data) => { do something with data here }
ftpConn.on('lookup', (data) => { do something with data here }

```
Since I'm using jsftp in a situation where the user supplies the hostname, I need to be able to give some useful response when the hostname is not valid.

In summary, there seem to be missing event handlers in JSFTP, and by adding them, and then adding similar handlers in my app code, I can prevent uncaught exception for host name lookup failures, or for ECONNREFUSED error due to an invalid port.

**Comment by mwittig** [comment]
@gabmontes The patches suggested by @davidjkrause are not required as the error event is propagated as it should. This is done by the event handler on the combined stream [line 141](https://github.com/sergi/jsftp/blob/master/lib/jsftp.js#L141). Note beyond, an "error" event will also be triggered on dns lookup failures. Thus the "lookup" event is not required here.

What you need to have is an "error" event handler as part of your code as @davidjkrause outlined, i.e.
```ftpConn.on('error', (data) => { do something with data here }```

In summarry, JSFTP does not lack event handlers, but the matter is more of an documentation issue.

**Comment by gabmontes** [comment]
@mwittig awesome! You are absolutely right. Adding the "error" event handler fixes the problem:

```js
const JSFtp = require('jsftp')

const ftp = new JSFtp({
  host: '127.0.0.1', port: 3333,
  user: 'user', pass: '1234'
})

ftp.ls('.', function(err, res) {
  console.log(res)
})

ftp.on('error', function (err) {
  console.log('Error captured: ', err.message)
})
```
